File: srclib/v3/R/residuals.R
# residuals function

residuals <- function() {
  grpvar <- ""  # build the r command
  estvar <- ""
  resvar <- ""
  genvars <- character()
  type <- ""
  if (grepl(".+\\.[[:alnum:]]+$", datain)) type <- tolower(sub(".+\\.([[:alnum:]]+)$", "\\1", datain))
  if (identical(type, "rda") || identical(type, "rdata")) {
    cmd <- paste("\nds <- get(load(\"", datain, "\"))\n", sep="")
  } else cmd <- paste("\nds <- get(load(\"", datain, ".rda\"))\n", sep="")
  cmd <- paste(cmd, "attach(ds)\n", sep="")
  res <- paste(path, "res", sep=".")
  f <- file(res, "r")  # read the residuals file
  while (length(line <- readLines(f, n = 1, warn = FALSE)) > 0) {
    if (identical(grpvar, "") && grepl("^\\s*[[:alnum:]]+\\s*=\\s*[[:digit:]]+\\s*;*\\s*$", line)) {
      grpvar <- sub("^\\s*([[:alnum:]]+)\\s*=\\s*[[:digit:]]+\\s*;*\\s*$", "\\1", line)
      grpval <- sub("^\\s*[[:alnum:]]+\\s*=\\s*([[:digit:]]+)\\s*;*\\s*$", "\\1", line)
      cmd <- paste(cmd, "ds$", grpvar, " <- ", grpval, "\n", sep="")
    }
    else if (grepl("^\\s*IF\\s+[[:alnum:]]+\\s+EQ\\s+[[:digit:]]+\\s*THEN*\\s*$", line, ignore.case=TRUE)) {
      grpvar <- sub("^\\s*IF\\s+([[:alnum:]]+)\\s+EQ\\s+[[:digit:]]+\\s*THEN*\\s*$", "\\1", line, ignore.case=TRUE)
      grpval <- sub("^\\s*IF\\s+[[:alnum:]]+\\s+EQ\\s+([[:digit:]]+)\\s*THEN*\\s*$", "\\1", line, ignore.case=TRUE)
      cmd <- paste(cmd, "ds$", grpvar, " <- ifelse(ds$", grpvar, " == ", grpval, ", ", sep="")
    }
    else if (grepl(paste("^\\s*IF\\s+[[:alnum:]]+\\s+IN\\([[:digit:],]+\\)\\s+THEN\\s+", grpvar, "\\s*=\\s*[[:digit:]]+\\s*;*\\s*$",
        sep=""), line, ignore.case=TRUE)) {
      spltvar <- sub(paste("^\\s*IF\\s+([[:alnum:]]+)\\s+IN\\([[:digit:],]+\\)\\s+THEN\\s+", grpvar, "\\s*=\\s*[[:digit:]]+\\s*;*\\s*$",
        sep=""), "\\1", line, ignore.case=TRUE)
      spltvals <- sub(paste("^\\s*IF\\s+[[:alnum:]]+\\s+IN\\(([[:digit:],]+)\\)\\s+THEN\\s+", grpvar, "\\s*=\\s*[[:digit:]]+\\s*;*\\s*$",
        sep=""), "\\1", line, ignore.case=TRUE)
      grpval <- sub(paste("^\\s*IF\\s+[[:alnum:]]+\\s+IN\\([[:digit:],]+\\)\\s+THEN\\s+", grpvar, "\\s*=\\s*([[:digit:]]+)\\s*;*\\s*$",
        sep=""), "\\1", line, ignore.case=TRUE)
      if (identical(grpval, "2")) {
        cmd <- paste(cmd, "ds$", grpvar, " <- ifelse(", spltvar, " %in% c(", spltvals, "), ", grpval, sep="")
      } else  cmd <- paste(cmd, "ifelse(", spltvar, " %in% c(", spltvals, "), ", grpval, sep="")
    }
    else if (grepl(paste("^\\s*ELSE\\s+", grpvar, "\\s*=\\s*[[:digit:]]+\\s*;*\\s*$", sep=""), line, ignore.case=TRUE)) {
      grpval <- sub(paste("^\\s*ELSE\\s+", grpvar, "\\s*=\\s*([[:digit:]]+)\\s*;*\\s*$", sep=""), "\\1", line, ignore.case=TRUE)
      if (identical(grpval, "3")) {
        cmd <- paste(cmd, ", ", grpval, ")\n", sep="")
      } else cmd <- paste(cmd, ", ", grpval, "), ds$", grpvar, ")\n", sep="")
      ifelseclose <- "))\n"
    }
    else if (grepl(paste("^\\s*IF\\s+", grpvar, "\\s+EQ\\s+[[:digit:]]+\\s+THEN\\s+[[:alnum:]]+\\s*=\\s*[[:alnum:]_\\.*+ -]+\\s*;*\\s*$", sep=""),
      line, ignore.case=TRUE)) {
      grpval <- sub(paste("^\\s*IF\\s+", grpvar, "\\s+EQ\\s+([[:digit:]]+)\\s+THEN\\s+[[:alnum:]]+\\s*=\\s*[[:alnum:]_\\.*+ -]+\\s*;*\\s*$", sep=""),
        "\\1", line, ignore.case=TRUE)
      estvar <- sub(paste("^\\s*IF\\s+", grpvar, "\\s+EQ\\s+[[:digit:]]+\\s+THEN\\s+([[:alnum:]]+)\\s*=\\s*[[:alnum:]_\\.*+ -]+\\s*;*\\s*$", sep=""),
        "\\1", line, ignore.case=TRUE)
      estexp <- sub(paste("^\\s*IF\\s+", grpvar, "\\s+EQ\\s+[[:digit:]]+\\s+THEN\\s+[[:alnum:]]+\\s*=\\s*([[:alnum:]_\\.*+ -]+)\\s*;*\\s*$", sep=""),
        "\\1", line, ignore.case=TRUE)
      if (!(estvar %in% genvars)) {
        genvars <- c(genvars, estvar)
        cmd <- paste(cmd, "ds$", estvar, " <- as.double(NA)\n", sep = "")
      }
      cmd <- paste(cmd, "ds$", estvar, " <- ", "ifelse(ds$", grpvar, " == ", grpval, ", ", estexp, ", ds$", estvar, ")\n", sep="")
    }
    else if (grepl("^\\s*[[:alnum:]]+\\s*=\\s*[[:alnum:]]+\\s*-\\s*[[:alnum:]]+\\s*;*\\s*$", line)) {
      resvar <- sub("^\\s*([[:alnum:]]+)\\s*=\\s*[[:alnum:]]+\\s*-\\s*[[:alnum:]]+\\s*;*\\s*$", "\\1", line)
      depvar <- sub("^\\s*[[:alnum:]]+\\s*=\\s*([[:alnum:]]+)\\s*-\\s*[[:alnum:]]+\\s*;*\\s*$", "\\1", line)
      estvar <- sub("^\\s*[[:alnum:]]+\\s*=\\s*[[:alnum:]]+\\s*-\\s*([[:alnum:]]+)\\s*;*\\s*$", "\\1", line)
      cmd <- paste(cmd, "ds$", resvar, " <- ", depvar, " - ds$", estvar, "\n", sep="")
    }
    else if (grepl(paste("^\\s*IF\\s+", grpvar, "\\s+EQ\\s+[[:digit:]]+\\s+THEN\\s+DO\\s*;*\\s*$", sep=""), line, ignore.case=TRUE)) {
      grpval <- sub(paste("^\\s*IF\\s+", grpvar, "\\s+EQ\\s+([[:digit:]]+)\\s+THEN\\s+DO\\s*;*\\s*$", sep=""), "\\1", line, ignore.case=TRUE)
    }
    else if (grepl("^\\s*[[:alnum:]]+\\s*=\\s*[[:digit:]\\.*+ -]+\\s*;*\\s*$", line)) {
      estvar <- sub("^\\s*([[:alnum:]]+)\\s*=\\s*[[:digit:]\\.*+ -]+\\s*;*\\s*$", "\\1", line)
      estexp <- sub("^\\s*[[:alnum:]]+\\s*=\\s*([[:digit:]\\.*+ -]+)\\s*;*\\s*$", "\\1", line)
      if (!(estvar %in% genvars)) {
        genvars <- c(genvars, estvar)
        cmd <- paste(cmd, "ds$", estvar, " <- as.double(NA)\n", sep = "")
      }
      cmd <- paste(cmd, "ds$", estvar, " <- ", "ifelse(ds$", grpvar, " == ", grpval, ", ", estexp, ", ds$", estvar, ")\n", sep="")
    }
    else if (grepl("^\\s*END\\s*;*\\s*$", line, ignore.case=TRUE)) {
    }
    else if (grepl("^\\s*IF\\s+[[:alnum:]]+\\s+EQ\\s+[[:digit:]]+\\s+THEN\\s+[[:alnum:]]+\\s*=\\s*[[:digit:]\\.*+ -]+\\s*;*\\s*$", line,
      ignore.case=TRUE)) {
      depvar <- sub("^\\s*IF\\s+([[:alnum:]]+)\\s+EQ\\s+[[:digit:]]+\\s+THEN\\s+[[:alnum:]]+\\s*=\\s*[[:digit:]\\.*+ -]+\\s*;*\\s*$", "\\1", line,
        ignore.case=TRUE)
      depval <- sub("^\\s*IF\\s+[[:alnum:]]+\\s+EQ\\s+([[:digit:]]+)\\s+THEN\\s+[[:alnum:]]+\\s*=\\s*[[:digit:]\\.*+ -]+\\s*;*\\s*$", "\\1", line,
        ignore.case=TRUE)
      resvar <- sub("^\\s*IF\\s+[[:alnum:]]+\\s+EQ\\s+[[:digit:]]+\\s+THEN\\s+([[:alnum:]]+)\\s*=\\s*[[:digit:]\\.*+ -]+\\s*;*\\s*$", "\\1", line,
        ignore.case=TRUE)
      resexp <- sub("^\\s*IF\\s+[[:alnum:]]+\\s+EQ\\s+[[:digit:]]+\\s+THEN\\s+[[:alnum:]]+\\s*=\\s*([[:digit:]\\.*+ -]+)\\s*;*\\s*$", "\\1", line,
        ignore.case=TRUE)
      if (!(resvar %in% genvars)) {
        genvars <- c(genvars, resvar)
        cmd <- paste(cmd, "ds$", resvar, " <- as.double(NA)\n", sep = "")
      }
      cmd <- paste(cmd, "ds$", resvar, " <- ", "ifelse(ds$", depvar, " == ", depval, ", ", resexp, ", ds$", resvar, ")\n", sep="")
    }
    else if (grepl("^\\s*IF\\s+[[:alnum:]]+\\s+EQ\\s+[[:digit:]\\.]+\\s+THEN\\s+[[:alnum:]]+\\s*=\\s*1\\s*-\\s*[[:alnum:]]+\\s*;*\\s*$", line,
      ignore.case=TRUE)) {
      depvar <- sub("^\\s*IF\\s+([[:alnum:]]+)\\s+EQ\\s+[[:digit:]\\.]+\\s+THEN\\s+[[:alnum:]]+\\s*=\\s*1\\s*-\\s*[[:alnum:]]+\\s*;*\\s*$", "\\1", line,
        ignore.case=TRUE)
      depval <- sub("^\\s*IF\\s+[[:alnum:]]+\\s+EQ\\s+([[:digit:]\\.]+)\\s+THEN\\s+[[:alnum:]]+\\s*=\\s*1\\s*-\\s*[[:alnum:]]+\\s*;*\\s*$", "\\1", line,
        ignore.case=TRUE)
      resvar <- sub("^\\s*IF\\s+[[:alnum:]]+\\s+EQ\\s+[[:digit:]\\.]+\\s+THEN\\s+([[:alnum:]]+)\\s*=\\s*1\\s*-\\s*[[:alnum:]]+\\s*;*\\s*$", "\\1", line,
        ignore.case=TRUE)
      if (identical(depval, ".")) {
        cmd <- paste(cmd, "ds$", resvar, " <- ", "ifelse(lapply(ds$", depvar, ", is.na), 1 - ds$", resvar, ", - ds$", resvar, ")\n", sep="")
      }
      else {
        cmd <- paste(cmd, "ds$", resvar, " <- ", "ifelse(lapply(ds$", depvar, ", is.na), - ds$", resvar, ", ifelse(ds$", depvar, " == ", depval, ", 1 - ds$", resvar,
          ", - ds$", resvar, "))\n", sep="")
      }
    }
    else if (grepl("^\\s*ELSE\\s+[[:alnum:]]+\\s*=\\s*[[:digit:]\\.*+ -]+\\s*;*\\s*$", line, ignore.case=TRUE)) {
      resvar <- sub("^\\s*ELSE\\s+([[:alnum:]]+)\\s*=\\s*[[:digit:]\\.*+ -]+\\s*;*\\s*$", "\\1", line, ignore.case=TRUE)
      resexp <- sub("^\\s*ELSE\\s+[[:alnum:]]+\\s*=\\s*([[:digit:]\\.*+ -]+)\\s*;*\\s*$", "\\1", line, ignore.case=TRUE)
      cmd <- paste(cmd, " else ", resvar, " <- ", resexp, "\n", sep="")
    }
    else if (grepl("^\\s*ELSE\\s+[[:alnum:]]+\\s*=\\s*-\\s*[[:alnum:]]+\\s*;*\\s*$", line, ignore.case=TRUE)) {
    }
    else {
      cat("Unrecognized line:", line, sep=" ")
      break
    }
  }
  close(f)
  type <- ""
  if (grepl(".+\\.[[:alnum:]]+$", dataout)) type <- tolower(sub(".+\\.([[:alnum:]]+)$", "\\1", dataout))
  if (identical(type, "rda") || identical(type, "rdata")) {
    cmd <- paste(cmd, "save(ds, file=\"", dataout, "\")\n", sep="")
  } else cmd <- paste(cmd, "save(ds, file=\"", dataout, ".rda\")\n\n", sep="")
  out <- paste(path, "out", sep=".")
  f <- file(out, "wb")  # write the residuals file
  writeChar(cmd, f, nchars = nchar(cmd, type = "chars"), eos = NULL)
  close(f)
  source(out)
}

File: srclib/v3/spss/srclib/residuals.py
# residuals function

import os, re, spss

def residuals(datain, dataout, path, mode):
  grpvar = ""  # build the spss command
  estvar = ""
  resvar = ""
  if not re.match(r"\.sav$", datain, flags=re.IGNORECASE):  # get datain
    datain += ".sav"  # add .sav
  cmd = "get file='{0}'.\n".format(datain)
  hold = 0
  with open(path + ".res", "r") as file:  # read the residuals file
    for line in file:
      if grpvar == "":
        m = re.match("^ *(\\w+) *= *(\\d+) *; *$", line)
        if m:
          grpvar = m.group(1)
          grpval = m.group(2)
          cmd += "compute {0} = {1}.\n".format(grpvar, grpval)
      else:
        m = re.match("^ *IF *{0} *EQ *(\\d+) *THEN *$".format(grpvar), line)
        if m:
          grpval = m.group(1)
          cmd += "do if ({0} eq {1}).\n".format(grpvar, grpval)
        else:
          m = re.match("^ *IF *([\\w]+) *IN\\(([\\d,]+)\\) *THEN *{0} *= *([\\d]+) *; *$".format(grpvar), line)
          if m:
            spltvar = m.group(1)
            spltvals = m.group(2)
            grpval = m.group(3)
            cmd += "recode {0} ({1} = {2})".format(spltvar, spltvals, grpval)
          else:
            m = re.match("^ *ELSE *{0} *= *([\\d]+) *; *$".format(grpvar), line)
            if m:
              grpval = m.group(1)
              cmd += " (else = {0}) into {1}.\n".format(grpval, grpvar)
              if grpval != "3":
                cmd += "end if.\n"
            else:
              m = re.match("^ *IF *{0} *EQ *([\\d]+) *THEN *([\\w]+) *= *(.+) *; *$".format(grpvar), line)
              if m:
                grpval = m.group(1)
                estvar = m.group(2)
                estexp = m.group(3)
                cmd += "if ({0} eq {1}) {2} = {3}.\n".format(grpvar, grpval, estvar, estexp)
              else:
                m = re.match("^ *([\\w]+) *= *([\\w]+) *\- *([\\w]+) *; *$")
                if m:
                  resvar = m.group(1)
                  depvar = m.group(2)
                  estvar = m.group(3)
                  if resvar == estvar:
                    cmd += "recode {0} = {1} - {2}.\n".format(resvar, depvar, estvar)
                  else:
                    cmd += "compute {0} = {1} - {2}.\n".format(resvar, depvar, estvar)
                else:
                  m = re.match("^ *IF *{0} *EQ *([\\d]+) *THEN *DO *; *$".format(grpvar), line)
                  if m:
                    grpval = m.group(1)
                    if hold == 0:
                      hold = 1
                      value = "do if ({0} eq {1}).\n".format(grpvar, grpval)
                    else:
                      if hold == 1:
                        hold = 2
                        cmd += value
                      cmd += "do if ({0} eq {1}).\n".format(grpvar, grpval)
                  else:
                    m = re.match("^ *([\\w]+) *= *(.+) *; *$")
                    if m:
                      estvar = m.group(1)
                      estexp = m.group(2)
                      if hold == 1:
                        cmd += "recode {0} = $sysmis.\n".format(estvar)
                        value += "recode {0} = {1}.\n".format(estvar, estexp)
                      else:
                        cmd += "recode {0} = {1}.\n".format(estvar, estexp)
                    else:
                      m = re.match("^ *IF *([\\w]+) *EQ *([\\d]+|.) *THEN *([\\w]+) *= *(.+) *; *$")
                      if m:
                        depvar = m.group(1)
                        depval = m.group(2)
                        resvar = m.group(3)
                        resexp = m.group(4)
                        cmd += "do if ({0} eq {1}).\nrecode {2} = {3}.\n".format(depvar, depval, resvar, resexp)
                      else:
                        m = re.match("^ *ELSE *([\\w]+) *= *(.+) *; *$")
                        if m:
                          resvar = m.group(1)
                          resexp = m.group(2)
                          cmd += "else.\nrecode {0} = {1}.\nend if.\n".format(resvar, resexp)
  if not re.match(r"\.sav$", dataout):
    dataout += ".sav"  # add .sav
  cmd += "save outfile='{0}'.\n".format(dataout)
  if mode:
    print cmd  # print the command
  spss.Submit(cmd)  # execute the command

